File: exercise/java/day03/src/test/java/PopulationTests.java
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import people.Person;
import people.Pet;
import people.PetType;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.lang.Integer.MAX_VALUE;
import static java.lang.String.format;
import static java.lang.System.lineSeparator;
import static org.assertj.core.api.Assertions.assertThat;

class PopulationTests {
    private static List<Person> population;

    @BeforeAll
    static void setup() {
        population = Arrays.asList(
                new Person("Peter", "Griffin")
                        .addPet(PetType.CAT, "Tabby", 2),
                new Person("Stewie", "Griffin")
                        .addPet(PetType.CAT, "Dolly", 3)
                        .addPet(PetType.DOG, "Brian", 9),
                new Person("Joe", "Swanson")
                        .addPet(PetType.DOG, "Spike", 4),
                new Person("Lois", "Griffin")
                        .addPet(PetType.SNAKE, "Serpy", 1),
                new Person("Meg", "Griffin")
                        .addPet(PetType.BIRD, "Tweety", 1),
                new Person("Chris", "Griffin")
                        .addPet(PetType.TURTLE, "Speedy", 4),
                new Person("Cleveland", "Brown")
                        .addPet(PetType.HAMSTER, "Fuzzy", 1)
                        .addPet(PetType.HAMSTER, "Wuzzy", 2),
                new Person("Glenn", "Quagmire")
        );
    }

    @Test
    void whoOwnsTheYoungestPet() {
        var filtered = population.stream().min(Comparator.comparingInt(person -> person.pets().stream().mapToInt(Pet::age).min().orElse(Integer.MAX_VALUE))).orElse(null);

        assert filtered != null;
        assertThat(filtered.firstName()).isEqualTo("Lois");
    }
}

File: exercise/kotlin/day03/src/test/kotlin/PopulationTests.kt
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import people.Person
import people.PetType.*

class PopulationTests : FunSpec({
    val population: List<Person> = listOf(
        Person("Peter", "Griffin")
            .addPet(CAT, "Tabby", 2),
        Person("Stewie", "Griffin")
            .addPet(CAT, "Dolly", 3)
            .addPet(DOG, "Brian", 9),
        Person("Joe", "Swanson")
            .addPet(DOG, "Spike", 4),
        Person("Lois", "Griffin")
            .addPet(SNAKE, "Serpy", 1),
        Person("Meg", "Griffin")
            .addPet(BIRD, "Tweety", 1),
        Person("Chris", "Griffin")
            .addPet(TURTLE, "Speedy", 4),
        Person("Cleveland", "Brown")
            .addPet(HAMSTER, "Fuzzy", 1)
            .addPet(HAMSTER, "Wuzzy", 2),
        Person("Glenn", "Quagmire")
    )

    test("who owns the youngest pet") {
        val filtered = population.filter { person -> person.pets.isNotEmpty() }
            .minBy { person -> person.pets.map { p -> p.age }.min() }
        filtered.firstName shouldBe "Lois"
    }
})
